### PR TITLE
Module 3 pull request [Rest api advanced]

### PR DESCRIPTION
**Pull request by YuKopilets**

### CODE REVIEW

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        return orders;
+    }
+
+    private void updateOrdersCost(List<Order> orders) {
+        orders.forEach(orderDao::update);
+    }
+
+    private void validateLogin(String login) {
+        int length = login.length();
+        if (!(length >= 4 && length <= 50)) {
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java:111]: Can you explain why we should use this values?

File: GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 2]
```diff
@@ -17,10 +18,9 @@
 
 import java.math.BigDecimal;
 import java.math.RoundingMode;
+import java.time.Duration;
 import java.time.LocalDateTime;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Optional;
+import java.util.*;
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:23]: **TBD:** Replace it with single import statements.

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/dto/GiftCertificatePatchDto.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * The {@code Gift certificate patch dto} is dto for transferring gift
+ * certificate data after request with <b>PATCH</b> http method.
+ *
+ * @author Yuriy Kopilets
+ * @version 1.0
+ */
+@Value
+public class GiftCertificatePatchDto {
+    @Size(min = 6, max = 255)
+    @Pattern(regexp = "^[A-Za-z0-9]+(\\s[A-Za-z0-9]+)*$")
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/dto/GiftCertificatePatchDto.java:22]: **TBD:** Does it make sense to simplify this regexp?

File: GiftCertificatesSystem/web/src/main/resources/application.properties [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+# DATASOURCE
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+spring.datasource.url=jdbc:mysql://localhost:3306/gift_certificates_system?useUnicode=true&serverTimezone=UTC&createDatabaseIfNotExist=true
+spring.datasource.username=root
+spring.datasource.password=root
+spring.datasource.hikari.maximum-pool-size=100
+
+# LOGGER
+#logging.level.root=debug
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/resources/application.properties:13]: Do we need this logging config?

**Review by maximelyashevich — CHANGES_REQUESTED** [review]

File: GiftCertificatesSystem/persistence/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java [hunk 5]
```diff
@@ -1,133 +1,106 @@
...
-        jdbcTemplate.update(UPDATE_GIFT_CERTIFICATE,
-                giftCertificate.getName(),
-                giftCertificate.getDescription(),
-                giftCertificate.getPrice(),
-                giftCertificate.getCreateDate(),
-                giftCertificate.getLastUpdateDate(),
-                giftCertificate.getDuration(),
-                giftCertificate.getId()
+        doWithSessionTransaction(session -> session.createQuery(UPDATE_GIFT_CERTIFICATE)
+                .setParameter("id", giftCertificate.getId())
```
[the remark this answers is not in the document]
Comment (YuKopilets) [reply GiftCertificatesSystem/persistence/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java:62]: Value of id substituted in codition part of query, not in SET.

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/dto/GiftCertificatePatchDto.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * @author Yuriy Kopilets
+ * @version 1.0
+ */
+@Getter
+public class GiftCertificatePatchDto {
+    private String name;
+    private String description;
+    private BigDecimal price;
+    private Integer duration;
+    private Set<Tag> tags;
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/dto/GiftCertificatePatchDto.java:22]: Can we modify tags?

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/dto/GiftCertificatePostDto.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    String description;
+
+    @NotNull
+    BigDecimal price;
+
+    @NotNull
+    Integer duration;
+
+    @NotEmpty
+    Set<Tag> tags;
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/dto/GiftCertificatePostDto.java:33]: Does it make sense to consider unmodifiable collection?

**Review by maximelyashevich — CHANGES_REQUESTED** [review]
Hi @YuKopilets,

Minor changes are required. And I couldn't find where you used JPA features to get audit data.

Regards,
Maks

File: GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 8]
```diff
@@ -116,57 +151,82 @@ void removeGiftCertificateNegativeTest() {
...
                 .id(1L)
                 .name("test name")
                 .description("test description")
                 .price(BigDecimal.valueOf(20.30).setScale(2, RoundingMode.HALF_UP))
                 .createDate(localDateTime)
                 .lastUpdateDate(localDateTime)
-                .duration(25)
+                .duration(Duration.ofDays(25))
+                .tags(tags)
                 .build();
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:165]: What's about `buildGiftCertificate()` method?

**Review by maximelyashevich — CHANGES_REQUESTED** [review]
Reviewed from my side, minor changes are required.

Regards,
Maks

File: GiftCertificatesSystem/persistence/src/main/java/com/epam/esm/entity/validator/DaysDurationValidator.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ * Validate that Duration holds value in positive number of days.
+ *
+ * @author Yuriy Kopilets
+ * @version 1.0
+ * @see ConstraintValidator
+ */
+public class DaysDurationValidator implements ConstraintValidator<ValidDaysDuration, Duration> {
+    @Override
+    public boolean isValid(Duration duration, ConstraintValidatorContext context) {
+        return duration.toDays() > 1;
```
[the remark this answers is not in the document]
Comment (YuKopilets) [reply GiftCertificatesSystem/persistence/src/main/java/com/epam/esm/entity/validator/DaysDurationValidator.java:18]: There is javadoc above class, where is typed about positive number of days. I read practice that we dont need write doc for overided method

[the remark this answers is not in the document]
Comment (maximelyashevich) [reply GiftCertificatesSystem/persistence/src/main/java/com/epam/esm/entity/validator/DaysDurationValidator.java:18]: I thought that positive numberscould be defined by `x > 0`.. But whatever's best for you.

**Review by maximelyashevich — APPROVED** [review]
Reviewed from my side, great job.

Regards,
Maks

**Pull request merged**